core/src/main/java/org/apache/iceberg/BaseFilesTable.java

```diff
@@ -223,34 +225,28 @@ ManifestFile manifest() {
   static class ContentFileStructWithMetrics implements StructLike {
     private final StructLike fileAsStruct;
     private final MetricsUtil.ReadableMetricsStruct readableMetrics;
-    private final int expectedSize;
+    private final int position;
 
     ContentFileStructWithMetrics(
-        int expectedSize,
-        StructLike fileAsStruct,
-        MetricsUtil.ReadableMetricsStruct readableMetrics) {
+        int position, StructLike fileAsStruct, MetricsUtil.ReadableMetricsStruct readableMetrics) {
       this.fileAsStruct = fileAsStruct;
       this.readableMetrics = readableMetrics;
-      this.expectedSize = expectedSize;
+      this.position = position;
     }
 
     @Override
     public int size() {
-      return expectedSize;
+      return position;
     }
 
     @Override
     public <T> T get(int pos, Class<T> javaClass) {
-      int lastExpectedIndex = expectedSize - 1;
-      if (pos < lastExpectedIndex) {
+      if (pos < position) {
         return fileAsStruct.get(pos, javaClass);
-      } else if (pos == lastExpectedIndex) {
+      } else if (pos == position) {
         return javaClass.cast(readableMetrics);
       } else {
-        throw new IllegalArgumentException(
```

Comment:
I don't quite understand the logic here. shouldn't we still need the `expectedSize` and throw an exception if the `pos` is over the size limit?

Reply:
I think we should just add a Preconditions check
```
Preconditions.check(pos < expectedSize, ...)
if (pos == metricsPosition) {
  return javaClass.cast...
} else {
 return  fileAsStruct.get(pos, javaClass);
}
```

Then just check if the (pos == metricsPos

Reply:
When pos is over the size limit, `fileAsStruct.get` method will throw an exception, we do not need to add Preconditions check, refer to the implementation of other `get` methods, I think this is the general way of it, unless we want to return a specific value, such as `null`, if it is over the limit.

Reply:
this else condition is essentially where `pos > metricsPosition`. why do we need to do minus ` of `fileAsStruct.get(pos - 1, javaClass);`? why not just `fileAsStruct.get(pos, javaClass);`

Isn't this logic simpler and sufficient?
```
if (pos == metricsPosition) {
  return javaClass.cast...
} else {
 return  fileAsStruct.get(pos, javaClass);
}
```

Reply:
metrics occupies a position.  `fileAsStruct` size + `Metrics` size(is 1) is equal to the size of `projection`. 

When pos is greater than metricsPosition, the actual position of the field in `fileAsStruct` should be subtracted by 1.
For instance:
```
fileAsStruct :[0->c1, 1->c2]
metrics :[0->c3]
project:[c1,c3,c2]
```
When we want to get `c2` (its pos is 2), but actually 1 in `fileAsStruct`.

Reply:
got it now. let's add some comments to explain the subtraction by 1.